contrib/opencensus-ext-azure/opencensus/ext/azure/metrics_exporter/__init__.py

```diff
@@ -12,30 +12,222 @@
 # limitations under the License.
 
-import threading
+import json
+import logging
+import requests
 
 from opencensus.ext.azure.common import Options
+from opencensus.ext.azure.common import utils
+from opencensus.ext.azure.common.protocol import Data
+from opencensus.ext.azure.common.protocol import DataPoint
+from opencensus.ext.azure.common.protocol import Envelope
+from opencensus.ext.azure.common.protocol import MetricData
 from opencensus.metrics import transport
+from opencensus.metrics.export.metric_descriptor import MetricDescriptorType
 from opencensus.stats import stats
+from opencensus.trace import execution_context
 
 __all__ = ['MetricsExporter', 'new_metrics_exporter']
 
+logger = logging.getLogger(__name__)
+
 
 class MetricsExporter(object):
     """Metrics exporter for Microsoft Azure Monitor."""
 
-    def __init__(self, options):
+    def __init__(self, options=None):
+        if options is None:
+            options = Options()
         self.options = options
         if not self.options.instrumentation_key:
             raise ValueError('The instrumentation_key is not provided.')
-        self._md_cache = {}
-        self._md_lock = threading.Lock()
+        self.max_batch_size = self.options.max_batch_size
 
     def export_metrics(self, metrics):
-        metrics = list(metrics)
-        for metric in metrics:
-            # TODO: implement export to Azure Monitor functionality
-            print(repr(metric))
+        if metrics:
+            envelopes = []
+            for metric in metrics:
+                # No support for histogram aggregations
+                type_ = metric.descriptor.type
+                if type_ != MetricDescriptorType.CUMULATIVE_DISTRIBUTION:
+                    md = metric.descriptor
+                    # Each time series will be uniquely
+                    # identified by it's label values
+                    for time_series in metric.time_series:
+                        # Using stats, time_series should
+                        # only have one point which contains
+                        # the aggregated value
+                        data_point = self.create_data_points(
+                            time_series, md)[0]
+                        # The timestamp is when the metric was recorded
+                        time_stamp = time_series.points[0].timestamp
+                        # Get the properties using label keys from metric
+                        # and label values of the time series
+                        properties = self.create_properties(time_series, md)
+                        envelopes.append(self.create_envelope(data_point,
+                                                              time_stamp,
+                                                              properties))
+                        # Send data in batches of max_batch_size
+                        if len(envelopes) == self.max_batch_size:
```

Comment:
If there are 10 metrics, each contains only one data point, do we end up sending 10 requests or just 1?

Reply:
If the max_batch_size value is 10,  one request is sent with a body serialized from a list with 10 envelopes.

Reply:
I see.